### PR TITLE
restore: skip unnecessary setgroups calls

### PR DESCRIPTION
**Pull request by angieenee**
When groups already have desired values, we should make a check to skip calling
sys_setgroup. Furthermore, this check allows us to restore without errors when in an
unprivileged user namespace that cannot use setgroups().

Signed-off-by: Angie Ni <avtni@google.com>

### CODE REVIEW

**Comment by kashyapp** [comment]
"restore: add check before calling sys_setgroups" happens to convey very little information. It just implies that some code around sys_setgroups is being changed.

The rest of the commit message looks good.

**Comment by kashyapp** [comment]
Though the original patch doesn't have it, this looks like a CL where it would be useful/interesting to add a test. @mclapinski do you agree?

**Comment by kashyapp** [comment]
See if there are existing tests in criu that cover setgroups behaviour.

**Comment by angieenee** [comment]
> "restore: add check before calling sys_setgroups" happens to convey very little information. It just implies that some code around sys_setgroups is being changed.
> 
> The rest of the commit message looks good.

100% agree. Thank you for pointing this out! I've just updated both the commit + PR message to be more specific to the intention of the changes.

File: criu/pie/restorer.c [hunk 1]
```diff
@@ -202,10 +202,21 @@ static int restore_creds(struct thread_creds_args *args, int procfd,
...
-			pr_err("Can't setup supplementary group IDs: %d\n", ret);
-			return -1;
+		// We may be in an unprivileged user namespace where setgroups
+		// is disabled.  If the current list of groups is already what
+		// we want, skip the call to setgroups.
+		unsigned int gids[32];
+		int n = sys_getgroups(32, gids);
+		// If gids is too small, n will be a negative error number so
+		// the comparison below will fail.
+		if (n == ce->n_groups && !memcmp(gids, args->groups, n * sizeof *gids)) {
```
Comment (angieenee) [inline criu/pie/restorer.c]: Alternative: sizeof(gids[0])

File: criu/pie/restorer.c [hunk 2]
```diff
@@ -202,10 +202,21 @@ static int restore_creds(struct thread_creds_args *args, int procfd,
...
 	if (args->groups) {
-		ret = sys_setgroups(ce->n_groups, args->groups);
-		if (ret) {
-			pr_err("Can't setup supplementary group IDs: %d\n", ret);
-			return -1;
+		// We may be in an unprivileged user namespace where setgroups
+		// is disabled.  If the current list of groups is already what
+		// we want, skip the call to setgroups.
+		unsigned int gids[32];
+		int n = sys_getgroups(32, gids);
```
Comment (angieenee) [inline criu/pie/restorer.c]: Call getgroups with 0 to get the size of the array, then allocate. Then call getgroups again with the actual size. 
int n = sys_getgroups(0, NULL);
// don't need to error check (might need to justify)
unsigned int* gids = malloc(n * sizeof(unsigned int)); // please brush up on C, Angie
// TODO: check if gids NULL
n = sys_getgroups(n, gids);

File: criu/pie/restorer.c [hunk 3]
```diff
@@ -202,10 +202,21 @@ static int restore_creds(struct thread_creds_args *args, int procfd,
...
+		// If gids is too small, n will be a negative error number so
+		// the comparison below will fail.
```
Comment (angieenee) [inline criu/pie/restorer.c]: Remove this comment because we ensure the size is correct earlier.

**Comment by angieenee** [comment]
[_copied from Files Changed tab for easier access_] Because restorer is in position-independent format, there are issues with dynamically allocating memory using malloc. That is likely why the original author of this patch wrote `unsigned int gids[32]` -- the original patch would still optimize this step in the restoring process if gids isn't too small.

I'm not 100% sure how to test out how efficient this optimization is, but I have isolated one test case that seems to be impacted by the `setgroups` call in this function. It'd be awesome to test how dramatic the optimization is and try out different gids sizes to find the best optimization while statically-allocating gids.

**Comment by mclapinski** [comment]
use sys_mmap to allocate memory instead of malloc

**Comment by kashyapp** [comment]
I have mislead slightly wrt the patch intent. Optimization is not the main goal, but to avoid the setgroups call as criu might be unprivileged to do so.

File: criu/pie/restorer.c [hunk 4]
```diff
@@ -202,11 +202,36 @@ static int restore_creds(struct thread_creds_args *args, int procfd,
...
+			if (ret) {
+				pr_err("Can't setgroups([%zu gids]): %d\n", ce->n_groups, ret);
+				ret = sys_munmap(gids, len);
+				if (ret) {
+					pr_err("Unable to deallocate %d bytes with munmap\n", len);
+				}
+				return -1;
+			}
+		}
+		ret = sys_munmap(gids, len);
```
Comment (kashyapp) [inline criu/pie/restorer.c]: See if the code can be restructured to avoid 2 calls to sys_munmap

File: criu/pie/restorer.c [hunk 5]
```diff
@@ -202,11 +202,31 @@ static int restore_creds(struct thread_creds_args *args, int procfd,
...
+			pr_debug("getgroups() looks good, not calling setgroups()\n");
+			ret = 0;
+		} else {
+			ret = sys_setgroups(ce->n_groups, args->groups);
+		}
+		if (ret)
+			pr_err("Can't setgroups([%zu gids]): %d\n", ce->n_groups, ret);
+		if (sys_munmap(gids, len))
+			pr_err("Unable to munmap %d bytes\n",len);
+		if (ret)
```
Comment (angieenee) [inline criu/pie/restorer.c]: I'm not sure how accepted it is to use `if (ret)` twice like this, but this is one way of avoiding a duplicate of `if (sys_munmap(gids, len))`

File: criu/pie/restorer.c [hunk 6]
```diff
@@ -202,11 +202,31 @@ static int restore_creds(struct thread_creds_args *args, int procfd,
...
 			return -1;
 		}
+		n = sys_getgroups(n, gids);
+		if (n == ce->n_groups && !memcmp(gids, args->groups, len)) {
+			pr_debug("getgroups() looks good, not calling setgroups()\n");
+			ret = 0;
+		} else {
+			ret = sys_setgroups(ce->n_groups, args->groups);
+		}
+		if (ret)
```
Comment (kashyapp) [inline criu/pie/restorer.c]: Shouldn't this be in the previous else {} block?

File: criu/pie/restorer.c (hunk 5 above)
Comment (kashyapp) [reply criu/pie/restorer.c]: I think that's fine, unlike errno, this is a function local.

File: criu/pie/restorer.c [hunk 7]
```diff
@@ -202,11 +202,31 @@ static int restore_creds(struct thread_creds_args *args, int procfd,
...
+			pr_err("Can't setgroups([%zu gids]): %d\n", ce->n_groups, ret);
+		if (sys_munmap(gids, len))
+			pr_err("Unable to munmap %d bytes\n",len);
```
Comment (kashyapp) [inline criu/pie/restorer.c]: The %d bytes part of this comment is unnecessary I think. The failure to unmap is unlikely to be due to size of the mapping.

File: criu/pie/restorer.c (hunk 5 above)
Comment (kashyapp) [reply criu/pie/restorer.c]: Maybe I shouldn't be the one marking as "resolved".

File: criu/pie/restorer.c (hunk 6 above)
Comment (angieenee) [reply criu/pie/restorer.c]: Ah yes! I'll fix that -- thank you.

File: criu/pie/restorer.c [hunk 8]
```diff
@@ -202,11 +202,31 @@ static int restore_creds(struct thread_creds_args *args, int procfd,
 	 * Setup supplementary group IDs early.
 	 */
 	if (args->groups) {
-		ret = sys_setgroups(ce->n_groups, args->groups);
-		if (ret) {
-			pr_err("Can't setup supplementary group IDs: %d\n", ret);
+		// We may be in an unprivileged user namespace where setgroups
```
Comment (mclapinski) [inline criu/pie/restorer.c]: change to "... where we're not allowed to call setgroups." and remove the double space after the full stop.

File: criu/pie/restorer.c [hunk 9]
```diff
@@ -202,11 +202,31 @@ static int restore_creds(struct thread_creds_args *args, int procfd,
...
 		}
+		n = sys_getgroups(n, gids);
+		if (n == ce->n_groups && !memcmp(gids, args->groups, len)) {
+			pr_debug("getgroups() looks good, not calling setgroups()\n");
+			ret = 0;
+		} else {
+			ret = sys_setgroups(ce->n_groups, args->groups);
+		}
+		if (ret)
+			pr_err("Can't setgroups([%zu gids]): %d\n", ce->n_groups, ret);
```
Comment (mclapinski) [inline criu/pie/restorer.c:225]: revert to `pr_err("Can't setup supplementary group IDs: %d\n", ret);`

**Review by kashyapp — CHANGES_REQUESTED** [review]
The commit message needs some re-wording

| restore: no sys_setgroups call if groups exist
this is wrong-ish, groups always exist

| When groups already have desired values, we should skip calling
| sys_setgroup. Therefore, another check is added in restore_creds to
| ensure auxiliary groups already containing desired values do not call
| sys_setgroup.
These two sentences seem to say the same thing.

| Furthermore, this check allows us to restore without errors when in an
| unprivileged user namespace that has sys_setgroups disabled.
Only when the groups don't need to be updated, else it fails.

I think the original commmit message is sufficient and concise.

Skip calling setgroups() when the list of auxiliary groups already has the
values we want.  This allows restoring into an unprivileged user namespace where
setgroups() is disabled.

**Comment by angieenee** [comment]
Thanks for the review, @kashyapp -- I've updated the commit name + description to better represent what this patch aims to achieve.

File: criu/pie/restorer.c [hunk 10]
```diff
@@ -202,11 +202,33 @@ static int restore_creds(struct thread_creds_args *args, int procfd,
...
+		groups_are_ok = (n == ce->n_groups && !memcmp(gids, args->groups, len));
+		if (sys_munmap(gids, len)) {
+			pr_err("Unable to munmap\n");
 			return -1;
 		}
+		if (groups_are_ok) {
+			pr_debug("getgroups() looks good, not calling setgroups()\n");
+		} else {
+			if (sys_setgroups(ce->n_groups, args->groups)) {
+				pr_err("Can't setgroups([%zu gids])\n", ce->n_groups);
```
Comment (mclapinski) [inline criu/pie/restorer.c:228]: please include errno in all 3 of the pr_err messages

File: criu/pie/restorer.c [hunk 11]
```diff
@@ -202,11 +202,33 @@ static int restore_creds(struct thread_creds_args *args, int procfd,
...
-			pr_err("Can't setup supplementary group IDs: %d\n", ret);
+		// We may be in an unprivileged user namespace where we're not
+		// allowed to call setgroups. If the current list of groups is already
+		// what we want, skip the call to setgroups.
+
+		int n = sys_getgroups(0, NULL);
+		int len = n * sizeof(unsigned int);
+		bool groups_are_ok;
+		unsigned int* gids = (unsigned int*)sys_mmap(NULL, len, PROT_WRITE | PROT_READ,
+       		MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
```
Comment (mclapinski) [inline criu/pie/restorer.c]: This line starts with spaces. I think you should replace them with a tab.

**Review by mclapinski — APPROVED** [review]
looks great!

**Pull request closed without merging**